preparation.py

```diff
@@ -28,55 +29,70 @@ def overlay_masks(data_dir, dataset, target_dir, category_ids, erode=0, is_small
-
-
-def overlay_mask_one_image(image_id, coco, category_ids, erode):
+        executor.map(_overlay_mask_one_image, image_ids)
+
+
+def overlay_mask_one_image(image_id, dataset, target_dir, coco, category_ids, erode, dilate, with_borders):
+    logger.info('Preparing image: {}'.format(image_id))
     image = coco.loadImgs(image_id)[0]
     image_size = (image["height"], image["width"])
     mask_overlayed = np.zeros(image_size).astype('uint8')
     distances = np.zeros(image_size)
     for category_nr, category_id in enumerate(category_ids):
-        if category_id != None:
+        if category_id is not None:
             annotation_ids = coco.getAnnIds(imgIds=image_id, catIds=[category_id, ])
             annotations = coco.loadAnns(annotation_ids)
+
+            if erode < 0 or dilate < 0:
+                raise ValueError('erode and dilate cannot be negative')
+
             if erode == 0:
                 mask, distances = overlay_masks_from_annotations(annotations=annotations,
                                                                  image_size=image_size,
                                                                  distances=distances)
-            elif erode > 0:
+            elif dilate == 0:
                 mask, _ = overlay_masks_from_annotations(annotations=annotations,
                                                          image_size=image_size)
                 mask_eroded, distances = overlay_eroded_masks_from_annotations(annotations=annotations,
                                                                                image_size=image_size,
-                                                                               area_percent=erode,
+                                                                               erode=erode,
                                                                                distances=distances)
                 mask = add_dropped_objects(mask, mask_eroded)
             else:
-                raise ValueError('erode cannot be negative')
+                mask, distances = overlay_eroded__dilated_masks_from_annotations(annotations=annotations,
+                                                                                 image_size=image_size,
+                                                                                 erode=erode,
+                                                                                 dilate=dilate,
+                                                                                 distances=distances)
             mask_overlayed = np.where(mask, category_nr, mask_overlayed)
 
     sizes = get_size_matrix(mask_overlayed)
-    distances = clean_distances(distances).astype(np.float16)
-    return mask_overlayed, distances, sizes, image
+    distances, distances_ = clean_distances(distances)
+
+    if with_borders:
+        borders = (distances_[:, :, 1] < 8 ) & (~mask_overlayed)
```

Comment:
@jakubczakon yes it is magic, but still should be parameterized. But again, like with small area threshold, I wonder if we treat is as a constant or a hyperparameter. After a second thought I believe they both should be defined in neptune.yaml, they are clearly parameters that we can experiment with and run different experiments to choose the best ones. Do you agree?

Reply:
@taraspiotr yes, especially since presenting the final solution will depend on those heavily. And it can be easily missed by someone (us in the future)